### PR TITLE
Various DX improvements

### PR DESCRIPTION
**Pull request by radex**
‼️ DO NOT MERGE - dependent branch ‼️

### CODE REVIEW

File: src/Query/test.js [hunk 1]
```diff
@@ -16,259 +16,285 @@ class MockTaskModel extends Model {
...
-    const query = new Query(mockCollection, [
-      Q.on('projects', 'team_id', 'abcdef'),
-      Q.where('left_column', 'right_value'),
-      Q.on('tag_assignments', 'tag_id', Q.oneOf(['a', 'b', 'c'])),
-    ])
-    expect(query.serialize()).toEqual({
-      table: 'mock_tasks',
-      description: query.description,
-      associations: query.associations,
+describe('Query', () => {
```
Comment (radex) [inline src/Query/test.js:19]: sorry for the massive diff - GitHub is not smart enough to see i only changed indentation - feel safe skipping until next comment

File: src/Query/test.js [hunk 2]
```diff
@@ -16,259 +16,285 @@ class MockTaskModel extends Model {
...
+      expect(query.pipe(wrap).wrapped).toBe(query)
+    })
   })
-  it('can pipe query', () => {
-    const query = new Query(mockCollection, [Q.on('projects', 'team_id', 'abcdef')])
-    const identity = a => a
-    expect(query.pipe(identity)).toBe(query)
-    const wrap = q => ({ wrapped: q })
-    expect(query.pipe(wrap).wrapped).toBe(query)
+  describe('fetching', () => {
```
Comment (radex) [inline src/Query/test.js:128]: this is the only section that I actually added/changed

File: src/Database/test.js [hunk 3]
```diff
@@ -4,619 +4,651 @@ import { noop } from '../utils/fp'
 import { CollectionChangeTypes } from '../Collection/common'
 import * as Q from '../QueryDescription'
 
-describe('unsafeResetDatabase', () => {
-  it('can reset database', async () => {
-    const { database, tasks } = mockDatabase({ actionsEnabled: true })
+describe('Database', () => {
+  it(`implements get()`, () => {
```
Comment (radex) [inline src/Database/test.js:8]: sorry about the massive diff- github's not smart enough to see it's just indentation. this test (next few lines) is the only thing I changed

File: src/Database/test.js [hunk 4]
```diff
@@ -4,619 +4,651 @@ import { noop } from '../utils/fp'
...
-      ),
-    )
-    const [task] = await tasks.query().fetch()
-    expect(task.name).toBe('foo1')
-  })
-})
+      expect(adapterBatchSpy).toHaveBeenCalledTimes(1)
+      expect(adapterBatchSpy).toHaveBeenLastCalledWith([['create', 'mock_tasks', model._raw]])
+    })
+    it(`can batch with an array passed as argument`, async () => {
```
Comment (radex) [inline src/Database/test.js:251]: and this

File: src/Database/test.js [hunk 5]
```diff
@@ -4,619 +4,651 @@ import { noop } from '../utils/fp'
...
+        database.batch(
+          tasks.prepareCreate(task => {
+            task.name = 'foo1'
+          }),
+        ),
+      )
+      const [task] = await tasks.query().fetch()
+      expect(task.name).toBe('foo1')
+    })
+    it(`throws an error if invalid arguments`, async () => {
```
Comment (radex) [inline src/Database/test.js:289]: and this

**Comment by radex** [comment]
@kokusGr CR please :) the diff looks scary, but it's mostly junk changes, the actual diff is small

‼️ DO NOT MERGE - dependent branch ‼️

File: src/Query/index.js [hunk 6]
```diff
@@ -81,6 +88,14 @@ export default class Query<Record: Model> {
     return toPromise(callback => this.collection._fetchQuery(this, callback))
   }
 
+  then<U>(
```
Comment (kokusGr) [inline src/Query/index.js:91]: that's cool :)

**Comment by kokusGr** [comment]
Nice
Zrzut ekranu 2020-06-18 o 11 41 26

File: src/Query/index.js (hunk 6 above)
Comment (radex) [reply src/Query/index.js:91]: @kokusGr sadly, `await tenable` does not type-check in Flow right now :( there's a pending PR for it. but hopefully soon enough.

**Pull request merged**